sdk/nodejs/runtime/callbacks.ts

```diff
@@ -0,0 +1,409 @@
+        this._server.forceShutdown();
+    }
+
+    private async invoke(
+        call: grpc.ServerUnaryCall<CallbackInvokeRequest, CallbackInvokeResponse>,
+        callback: grpc.sendUnaryData<CallbackInvokeResponse>,
+    ) {
+        const req = call.request;
+
+        const cb = this._callbacks.get(req.getToken());
+        if (cb === undefined) {
+            const err = new grpc.StatusBuilder();
+            err.withCode(grpc.status.INVALID_ARGUMENT);
+            err.withDetails("callback not found: " + req.getToken());
+            callback(err.build());
+            return;
+        }
+
+        try {
+            const response = await cb(req.getRequest_asU8());
+            const resp = new CallbackInvokeResponse();
+            resp.setResponse(response.serializeBinary());
+            callback(null, resp);
+        } catch (e) {
+            const err = new grpc.StatusBuilder();
+            err.withCode(grpc.status.UNKNOWN);
+            if (e instanceof Error) {
+                err.withDetails(e.message);
+            } else {
+                err.withDetails(JSON.stringify(e));
+            }
+            callback(err.build());
+        }
+    }
+
+    async registerTransformation(transform: ResourceTransform): Promise<callproto.Callback> {
+        const cb = async (bytes: Uint8Array): Promise<jspb.Message> => {
+            const request = resproto.TransformRequest.deserializeBinary(bytes);
+
+            let opts = request.getOptions() || new resproto.TransformResourceOptions();
+
+            let ropts: ResourceOptions;
+            if (request.getCustom()) {
+                ropts = {
+                    deleteBeforeReplace: opts.getDeleteBeforeReplace(),
+                    additionalSecretOutputs: opts.getAdditionalSecretOutputsList(),
+                } as CustomResourceOptions;
+            } else {
+                const providers: Record<string, ProviderResource> = {};
+                for (const [key, value] of opts.getProvidersMap().entries()) {
+                    providers[key] = new DependencyProviderResource(value);
+                }
+                ropts = {
+                    providers: providers,
+                } as ComponentResourceOptions;
+            }
+
+            ropts.aliases = opts.getAliasesList().map((alias): string | Alias => {
+                if (alias.hasUrn()) {
+                    return alias.getUrn();
+                } else {
+                    const spec = alias.getSpec();
+                    if (spec === undefined) {
+                        throw new Error("alias must have either a urn or a spec");
+                    }
+                    return {
+                        name: spec.getName(),
+                        type: spec.getType(),
+                        project: spec.getProject(),
+                        stack: spec.getStack(),
+                        parent: spec.getParenturn() !== "" ? new DependencyResource(spec.getParenturn()) : undefined,
```

Comment:
I'd have to look closer, but is this right? Should we be checking the spec's `noParent`?

Reply:
Yup we did need to check that